plugin/evm/vm_warp_test.go

```diff
@@ -0,0 +1,338 @@
+	"github.com/ava-labs/subnet-evm/core"
+	"github.com/ava-labs/subnet-evm/core/rawdb"
+	"github.com/ava-labs/subnet-evm/core/types"
+	"github.com/ava-labs/subnet-evm/internal/ethapi"
+	"github.com/ava-labs/subnet-evm/params"
+	"github.com/ava-labs/subnet-evm/rpc"
+	predicateutils "github.com/ava-labs/subnet-evm/utils/predicate"
+	warpPayload "github.com/ava-labs/subnet-evm/warp/payload"
+	"github.com/ava-labs/subnet-evm/x/warp"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/common/hexutil"
+	"github.com/stretchr/testify/require"
+)
+
+func TestSendWarpMessage(t *testing.T) {
+	genesis := &core.Genesis{}
+	if err := genesis.UnmarshalJSON([]byte(genesisJSONSubnetEVM)); err != nil {
+		t.Fatal(err)
+	}
+	genesis.Config.GenesisPrecompiles = params.Precompiles{
+		warp.ConfigKey: warp.NewDefaultConfig(big.NewInt(0)),
+	}
+	genesisJSON, err := genesis.MarshalJSON()
+	if err != nil {
+		t.Fatal(err)
+	}
+	issuer, vm, _, _ := GenesisVM(t, true, string(genesisJSON), "", "")
+
+	defer func() {
+		if err := vm.Shutdown(context.Background()); err != nil {
+			t.Fatal(err)
+		}
+	}()
+
+	newTxPoolHeadChan := make(chan core.NewTxPoolReorgEvent, 1)
+	reorgSub := vm.txPool.SubscribeNewReorgEvent(newTxPoolHeadChan)
+	defer reorgSub.Unsubscribe()
+
+	acceptedLogsChan := make(chan []*types.Log, 10)
+	logsSub := vm.eth.APIBackend.SubscribeAcceptedLogsEvent(acceptedLogsChan)
+	defer logsSub.Unsubscribe()
+
+	payload := utils.RandomBytes(100)
+
+	warpSendMessageInput, err := warp.PackSendWarpMessage(warp.SendWarpMessageInput{
+		DestinationChainID: vm.ctx.CChainID,
+		DestinationAddress: testEthAddrs[1].Hash(),
+		Payload:            payload,
+	})
+	require.NoError(t, err)
+
+	// Submit a transaction to trigger sending a warp message
+	tx0 := types.NewTransaction(uint64(0), warp.ContractAddress, big.NewInt(1), 100_000, big.NewInt(testMinGasPrice), warpSendMessageInput)
+	signedTx0, err := types.SignTx(tx0, types.LatestSignerForChainID(vm.chainConfig.ChainID), testKeys[0])
+	require.NoError(t, err)
+
+	errs := vm.txPool.AddRemotesSync([]*types.Transaction{signedTx0})
+	if err := errs[0]; err != nil {
+		t.Fatalf("Failed to add tx at index: %s", err)
+	}
+
+	<-issuer
+	blk, err := vm.BuildBlock(context.Background())
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	if err := blk.Verify(context.Background()); err != nil {
+		t.Fatal(err)
+	}
+
+	if status := blk.Status(); status != choices.Processing {
+		t.Fatalf("Expected status of built block to be %s, but found %s", choices.Processing, status)
+	}
+
+	// Verify that the constructed block contains the expected log with an unsigned warp message in the log data
+	ethBlock1 := blk.(*chain.BlockWrapper).Block.(*Block).ethBlock
+	require.Len(t, ethBlock1.Transactions(), 1)
+	receipts := rawdb.ReadReceipts(vm.chaindb, ethBlock1.Hash(), ethBlock1.NumberU64(), vm.chainConfig)
+	require.NotNil(t, receipts)
+	require.Len(t, receipts, 1)
+
+	logData := receipts[0].Logs[0].Data
```

Comment:
can we also verify the topics are as expected?

Reply:
Updated